### PR TITLE
Recursive search for `.env` files

### PR DESCRIPTION
**Pull request by DDtKey**
I think it is useful that the linter can search for all `env` files in the specified directory, including the subdirectories.
This does not prevent specifying specific files and directories input, as well as excluding them.

What do you think?

#### ✔ Checklist:

- [X] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features).

### CODE REVIEW

**Comment by DDtKey** [comment]
I found #87 issue, it's already discussed, but in fact, I'm not sure what to compare the linter with [libraries for load environment variables](https://github.com/dotenv-linter/dotenv-linter/issues/87#issuecomment-590050476) is correctly.

Most of the linters that I know by default still know how to work recursively.
We can add a flag to turn off recursion or turn it on, as an option, but it seems to me that this approach is more suitable for a linter.

I will be glad to discuss this and look at it from another perspective.

**Comment by mstruebing** [comment]
Like in #87 my mind didn't change.
I think for linters to be most usable they should check recursively but if there are valid reasons against I don't know.
One thing we wouldn't can do as other linters mostly do is to only check version control system tracked files, because `.env`-files usually are not version tracked.

**Comment by mgrachev** [comment]
@DDtKey Thank you!

It would be great if we first discussed this feature before its implementation 🙏
Because, I don't see any reasons to do recursive search for `.env` files.

I have o lot of projects with `.env` files (several files per project) and I haven't had situations where I need to use recursive search.

Other linters like `rubocop` can search recursively because projects written in Ruby can have subdirectories with `.rb` files.

To use `.env` files in our projects we should use `dotenv` libraries whick look for `.env` files in the current directory:
* Ruby: https://github.com/bkeepers/dotenv
* Rust: https://github.com/dotenv-rs/dotenv
* JS: https://github.com/motdotla/dotenv
* PHP: https://github.com/vlucas/phpdotenv
* Haskell: https://github.com/stackbuilders/dotenv-hs
* Python: https://github.com/theskumar/python-dotenv

Or we should use Docker Compose and keep `.env` files somewhere else (if we want it).
I use both of ways every day and don't need to use recursive search.

**Comment by DDtKey** [comment]
> @DDtKey Thank you!
> 
> It would be great if we first discussed this feature before its implementation pray
> Because, I don't see any reasons to do recursive search for `.env` files.
> 
> I have o lot of projects with `.env` files (several files per project) and I haven't had situations where I need to use recursive search.
> 
> Other linters like `rubocop` can search recursively because projects written in Ruby can have subdirectories with `.rb` files.
> 
> To use `.env` files in our projects we should use `dotenv` libraries whick look for `.env` files in the current directory:
> 
>     * Ruby: https://github.com/bkeepers/dotenv
> 
>     * Rust: https://github.com/dotenv-rs/dotenv
> 
>     * JS: https://github.com/motdotla/dotenv
> 
>     * PHP: https://github.com/vlucas/phpdotenv
> 
>     * Haskell: https://github.com/stackbuilders/dotenv-hs
> 
>     * Python: https://github.com/theskumar/python-dotenv
> 
> 
> Or we should use Docker Compose and keep `.env` files somewhere else (if we want it).
> I use both of ways every day and don't need to use recursive search.

Yes, my fault :)

But as I wrote above, **loading variables is not exactly the same as a linter**.

In my memory, there were cases when folders were grouped and had other files nested. And they were directly loaded by applications according to completely different scenarios.

For example, take `clippy`, because we can write many modules, and it will check everything, regardless of usage. Most linter do so.

In the end, I just do not see a problem in the presence of this function, it allows you to do the same thing, but also complements the feature :thinking:

**Comment by DDtKey** [comment]
If this seems like an excess, okay, I will close the PR.
Thanks for the review!

**Comment by mgrachev** [comment]
> Yes, my fault :)
> 
> But as I wrote above, **loading variables is not exactly the same as a linter**.
> 
> In my memory, there were cases when folders were grouped and had other files nested. And they were directly loaded by applications according to completely different scenarios.

Usually all `.env` files are kept in the root directory of projects or in another directory.
Right now we can check these files with the command:
```sh
$ dotenv-linter ./another-directory .
```

Do you know projects where `.env` files are stored differently?
It would be nice if you gave an example.

> For example, take `clippy`, because we can write many modules, and it will check everything, regardless of usage. Most linter do so.

I think, `clippy` is a bad example, because it works with projects written in Rust. In Rust projects we can have such structure:
```sh
├── src
│   ├── checks
│   │   ├── duplicated_key.rs
│   │   ├── ending_blank_line.rs
│   │   ├── extra_blank_line.rs
│   │   ├── incorrect_delimiter.rs
│   │   ├── key_without_value.rs
│   │   ├── leading_character.rs
│   │   ├── lowercase_key.rs
│   │   ├── quote_character.rs
│   │   ├── space_character.rs
│   │   ├── trailing_whitespace.rs
│   │   └── unordered_key.rs
│   ├── checks.rs
│   ├── common.rs
│   ├── fs_utils.rs
│   ├── lib.rs
│   └── main.rs
``` 

But, I haven't seen any projects with similar structure with `.env` files:
```sh
├── example.env
├── global.env
├── subdir1
│   ├── common.env
│   └── subdir3
│       └── sample.env
└── subdir2
    └── test.env
```

> In the end, I just do not see a problem in the presence of this function, it allows you to do the same thing, but also complements the feature 🤔

I see some problems:
1. Recursive search is slower than search in only one directory;
2. I know a lot of projects that store their dependecies in some directories (`vendor`, `node_modules`) and after recursive search we may get some warnings from dependecies or their dependecies.

**Comment by mgrachev** [comment]
> If this seems like an excess, okay, I will close the PR.
> Thanks for the review!

I think we can add this feature, but only as an additional option.

@mstruebing What do you think about it?

**Comment by DDtKey** [comment]
I really saw projects with the following structure :
```sh
environments
├── config.env
├── module1
│   ├── module.env
└── module2
    └── module2.env
```
But with a deep level of nesting, probably not, I have not seen.

I'm not saying that this is a best practice. But this happens and maybe someone is useful.

**Comment by DDtKey** [comment]
>     2. I know a lot of projects that store their dependecies in some directories (`vendor`, `node_modules`) and after recursive search we may get some warnings from dependecies or their dependecies.

Such cases can always be handled with exclude (it works well in the current PR), but yes, we can to turn off recursion by default.

**Comment by wesleimp** [comment]
In my experience, like @DDtKey, I have seen many projects with several `.env` files, mainly when monorepo (e.g. `lerna`) is used for frontend. I think this function is very useful and is nice to have this feature.

About the implementation, when recursion is used, will there be any default exclude list (`node_modules`, `vendor`, etc)?

**Comment by mstruebing** [comment]
I think adding it as an optional feature is okay.
I currently work in a monorepo for example where every sub-package hat it's own `.env`-file like @DDtKey already mentioned, for that case it would be useful.
But mostly it's not an issue as most projects have a single `.env` file at root level.

So yeah, I agree: optional feature :+1:

**Comment by DDtKey** [comment]
@mgrachev Before doing any further actions for an optional feature, I would like to know your opinion.
If this is still useful, I will definitely make a change :relaxed:

**Comment by mgrachev** [comment]
I agree with @mstruebing. Please, add an additional flag to run recursive search.

**Comment by DDtKey** [comment]
> I agree with @mstruebing. Please, add an additional flag to run recursive search.

Added a flag, updated tests and documentation.
Recursive search only works when a flag is specified.

Glad to discuss possible codebase improvements :relaxed:

File: tests/args/specific_path.rs [hunk 1]
```diff
@@ -93,11 +93,11 @@ fn checks_one_specific_file_and_one_path() {
 
     let args = &[testfile_2.as_str(), subdir.as_str()];
     let expected_output = format!(
-        "{}:2 DuplicatedKey: The FOO key is duplicated\n{}:2 UnorderedKey: The BAR key should go before the FOO key\n",
+        "{}:2 UnorderedKey: The BAR key should go before the FOO key\n{}:2 DuplicatedKey: The FOO key is duplicated\n",
```
[the remark this answers is not in the document]
Comment (DDtKey) [reply tests/args/specific_path.rs:96]: The order was not entirely correct, it was more reasonable to display the directory files after the specified files.

File: src/lib.rs [hunk 2]
```diff
@@ -78,6 +60,38 @@ pub fn run(args: clap::ArgMatches, current_dir: &PathBuf) -> Result<Vec<Warning>
...
+                .collect()
+        })
+        .flat_map(|dir_entries| get_file_paths(dir_entries, excludes, recursive))
+        .collect();
+
+    let mut file_paths: Vec<PathBuf> = dir_entries
+        .into_iter()
+        .filter(|entry| entry.is_file())
+        .filter(|entry| !excludes.contains(entry))
+        .collect();
```
[the remark this answers is not in the document]
Comment (DDtKey) [reply src/lib.rs:87]: > What about cheching that this entry is `.env` file?

We check them when reading directories, excluding any other files.
At the same time, we allow you to specify a specific path to the file (which [was previously possible](https://github.com/dotenv-linter/dotenv-linter/blob/683988b3e276c17cc334b63a2cd203fb2df5db20/tests/args/specific_path.rs#L53))

File: README.md [hunk 3]
```diff
@@ -169,22 +169,30 @@ dir1/.env:3 IncorrectDelimiter: The FOO-BAR key has incorrect delimiter
...
 .env:2 DuplicatedKey: The FOO key is duplicated
 .env:3 UnorderedKey: The BAR key should go before the FOO key
 ```
 
+If you need a recursive `.env` file search inside directories, you can use the flag `--recursive` or its short version `-r`:
+
+```shell script
+$ dotenv-linter --recursive
+./dir1/.env:2 DuplicatedKey: The FOO key is duplicated
+./dir2/subdir/.env:3 IncorrectDelimiter: The FOO-BAR key has incorrect delimiter
```
[the remark this answers is not in the document]
Comment (DDtKey) [reply README.md:185]: No, good catch)

**Comment by mgrachev** [comment]
@mstruebing Please, take a look at these changes 👀

**Pull request merged**